File: src/events/types/git.ts
/**
 * Git 활동 이벤트 타입 정의
 */

import { BaseEvent, EventCategory, EventSeverity, EventMetadata } from './base.js';

/**
 * Git 이벤트 타입
 */
export enum GitEventType {
  // 커밋 관련
  COMMIT_CREATED = 'git:commit:created',
  COMMIT_AMENDED = 'git:commit:amended',

  // 브랜치 관련
  BRANCH_CREATED = 'git:branch:created',
  BRANCH_DELETED = 'git:branch:deleted',
  BRANCH_SWITCHED = 'git:branch:switched',
  BRANCH_MERGED = 'git:branch:merged',
  BRANCH_REBASED = 'git:branch:rebased',

  // 태그 관련
  TAG_CREATED = 'git:tag:created',
  TAG_DELETED = 'git:tag:deleted',

  // 원격 저장소 관련
  PUSH = 'git:push',
  PULL = 'git:pull',
  FETCH = 'git:fetch',
  CLONE = 'git:clone',

  // 병합 관련
  MERGE_STARTED = 'git:merge:started',
  MERGE_COMPLETED = 'git:merge:completed',
  MERGE_CONFLICT = 'git:merge:conflict',
  MERGE_ABORTED = 'git:merge:aborted',

  // Stash 관련
  STASH_CREATED = 'git:stash:created',
  STASH_APPLIED = 'git:stash:applied',
  STASH_DROPPED = 'git:stash:dropped',

  // 리베이스 관련
  REBASE_STARTED = 'git:rebase:started',
  REBASE_COMPLETED = 'git:rebase:completed',
  REBASE_CONFLICT = 'git:rebase:conflict',
  REBASE_ABORTED = 'git:rebase:aborted',
}

/**
 * Git 커밋 정보
 */
export interface GitCommitInfo {
  /** 커밋 해시 */
  hash: string;

  /** 짧은 해시 */
  shortHash: string;

  /** 커밋 메시지 */
  message: string;

  /** 작성자 */
  author: {
    name: string;
    email: string;
    date: Date;
  };

  /** 커밋터 */
  committer?: {
    name: string;
    email: string;
    date: Date;
  };

  /** 부모 커밋 해시들 */
  parents: string[];

  /** 변경된 파일 수 */
  filesChanged?: number;

  /** 추가된 줄 수 */
  insertions?: number;

  /** 삭제된 줄 수 */
  deletions?: number;

  /** 커밋 타입 (Conventional Commits) */
  type?: string;

  /** 커밋 스코프 */
  scope?: string;

  /** Breaking change 여부 */
  breaking?: boolean;
}

/**
 * Git 브랜치 정보
 */
export interface GitBranchInfo {
  /** 브랜치 이름 */
  name: string;

  /** 원격 브랜치 여부 */
  isRemote: boolean;

  /** 현재 브랜치 여부 */
  isCurrent: boolean;

  /** 업스트림 브랜치 */
  upstream?: string;

  /** 마지막 커밋 해시 */
  lastCommitHash?: string;

  /** 마지막 커밋 메시지 */
  lastCommitMessage?: string;

  /** 마지막 커밋 날짜 */
  lastCommitDate?: Date;

  /** ahead/behind 정보 */
  tracking?: {
    ahead: number;
    behind: number;
  };
}

/**
 * Git 태그 정보
 */
export interface GitTagInfo {
  /** 태그 이름 */
  name: string;

  /** 태그 타입 (lightweight/annotated) */
  type: 'lightweight' | 'annotated';

  /** 태그가 가리키는 커밋 해시 */
  commitHash: string;

  /** 태그 메시지 (annotated 태그의 경우) */
  message?: string;

  /** 태거 정보 */
  tagger?: {
    name: string;
    email: string;
    date: Date;
  };
}

/**
 * Git 원격 저장소 정보
 */
export interface GitRemoteInfo {
  /** 원격 저장소 이름 */
  name: string;

  /** URL */
  url: string;

  /** 타입 (fetch/push) */
  type: 'fetch' | 'push';
}

/**
 * Git 충돌 정보
 */
export interface GitConflictInfo {
  /** 충돌 파일들 */
  files: string[];

  /** 충돌 타입 */
  type: 'merge' | 'rebase' | 'cherry-pick';

  /** 소스 브랜치 */
  sourceBranch?: string;

  /** 타겟 브랜치 */
  targetBranch?: string;

  /** 충돌 상세 정보 */
  details?: Array<{
    file: string;
    conflictType:
      | 'both-modified'
      | 'deleted-by-them'
      | 'deleted-by-us'
      | 'added-by-them'
      | 'added-by-us';
  }>;
}

/**
 * Git 커밋 이벤트 데이터
 */
export interface GitCommitEventData {
  /** 커밋 정보 */
  commit: GitCommitInfo;

  /** 브랜치 정보 */
  branch: GitBranchInfo;

  /** 변경된 파일들 */
  files?: Array<{
    path: string;
    status: 'added' | 'modified' | 'deleted' | 'renamed';
    additions: number;
    deletions: number;
  }>;

  /** 커밋 패턴 분석 */
  analysis?: {
    /** 커밋 타입 분포 */
    isFeature?: boolean;
    isBugfix?: boolean;
    isRefactor?: boolean;
    isChore?: boolean;
    isDocs?: boolean;
    isTest?: boolean;
  };
}

/**
 * Git 브랜치 이벤트 데이터
 */
export interface GitBranchEventData {
  /** 브랜치 정보 */
  branch: GitBranchInfo;

  /** 이전 브랜치 (switch의 경우) */
  previousBranch?: string;

  /** 소스 브랜치 (생성/머지의 경우) */
  sourceBranch?: string;

  /** 병합 정보 (머지의 경우) */
  mergeInfo?: {
    commits: number;
    files: number;
    strategy: 'fast-forward' | 'recursive' | 'octopus' | 'ours' | 'subtree';
  };
}

/**
 * Git 푸시/풀 이벤트 데이터
 */
export interface GitSyncEventData {
  /** 원격 저장소 */
  remote: GitRemoteInfo;

  /** 브랜치 */
  branch: string;

  /** 동기화된 커밋 수 */
  commits?: number;

  /** 전송된 바이트 */
  bytes?: number;

  /** 동기화 방향 */
  direction: 'push' | 'pull' | 'fetch';

  /** 성공 여부 */
  success: boolean;

  /** 에러 메시지 */
  error?: string;
}

/**
 * Git 이벤트
 */
export interface GitEvent extends BaseEvent {
  type: GitEventType;
  category: EventCategory.GIT;
  data: GitCommitEventData | GitBranchEventData | GitSyncEventData | GitConflictInfo;
}

/**
 * Git 이벤트 생성 헬퍼
 */
export class GitEventBuilder {
  static createCommitEvent(
    type: GitEventType,
    data: GitCommitEventData,
    options?: {
      severity?: EventSeverity;
      correlationId?: string;
      metadata?: Record<string, any>;
    },
  ): GitEvent {
    return {
      id: `git-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`,
      type,
      category: EventCategory.GIT,
      timestamp: Date.now(),
      severity: options?.severity || EventSeverity.INFO,
      source: 'GitMonitor',
      data,
      ...(options?.correlationId && { correlationId: options.correlationId }),
      ...(options?.metadata && { metadata: options.metadata as EventMetadata }),
    };
  }

  static createBranchEvent(
    type: GitEventType,
    data: GitBranchEventData,
    options?: {
      severity?: EventSeverity;
      correlationId?: string;
      metadata?: Record<string, any>;
    },
  ): GitEvent {
    return {
      id: `git-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`,
      type,
      category: EventCategory.GIT,
      timestamp: Date.now(),
      severity: options?.severity || EventSeverity.INFO,
      source: 'GitMonitor',
      data,
      ...(options?.correlationId && { correlationId: options.correlationId }),
      ...(options?.metadata && { metadata: options.metadata as EventMetadata }),
    };
  }

  static createSyncEvent(
    type: GitEventType,
    data: GitSyncEventData,
    options?: {
      severity?: EventSeverity;
      correlationId?: string;
      metadata?: Record<string, any>;
    },
  ): GitEvent {
    return {
      id: `git-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`,
      type,
      category: EventCategory.GIT,
      timestamp: Date.now(),
      severity: data.success ? EventSeverity.INFO : EventSeverity.ERROR,
      source: 'GitMonitor',
      data,
      ...(options?.correlationId && { correlationId: options.correlationId }),
      ...(options?.metadata && { metadata: options.metadata as EventMetadata }),
    };
  }

  static createConflictEvent(
    type: GitEventType,
    data: GitConflictInfo,
    options?: {
      correlationId?: string;
      metadata?: Record<string, any>;
    },
  ): GitEvent {
    return {
      id: `git-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`,
      type,
      category: EventCategory.GIT,
      timestamp: Date.now(),
      severity: EventSeverity.WARNING,
      source: 'GitMonitor',
      data,
      ...(options?.correlationId && { correlationId: options.correlationId }),
      ...(options?.metadata && { metadata: options.metadata as EventMetadata }),
    };
  }
}

/**
 * Git 이벤트 타입 가드
 */
export function isGitEvent(event: BaseEvent): event is GitEvent {
  return event.category === EventCategory.GIT;
}

export function isGitCommitEvent(event: BaseEvent): event is GitEvent {
  return (
    event.category === EventCategory.GIT &&
    (event.type === GitEventType.COMMIT_CREATED || event.type === GitEventType.COMMIT_AMENDED)
  );
}

export function isGitBranchEvent(event: BaseEvent): event is GitEvent {
  return event.category === EventCategory.GIT && event.type.startsWith('git:branch:');
}

export function isGitSyncEvent(event: BaseEvent): event is GitEvent {
  return (
    event.category === EventCategory.GIT &&
    (event.type === GitEventType.PUSH ||
      event.type === GitEventType.PULL ||
      event.type === GitEventType.FETCH ||
      event.type === GitEventType.CLONE)
  );
}

File: python/src/devflow_monitor/events/types/git.py
"""
Git Activity Event Types.

This module defines event types for Git operations including
commits, branches, merges, pushes, pulls, and conflicts.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field

from .base import BaseEvent, EventCategory, EventMetadata, EventSeverity


class GitEventType(str, Enum):
    """Git event types."""

    # Commit related
    COMMIT_CREATED = "git:commit:created"
    COMMIT_AMENDED = "git:commit:amended"

    # Branch related
    BRANCH_CREATED = "git:branch:created"
    BRANCH_DELETED = "git:branch:deleted"
    BRANCH_SWITCHED = "git:branch:switched"
    BRANCH_MERGED = "git:branch:merged"
    BRANCH_REBASED = "git:branch:rebased"

    # Tag related
    TAG_CREATED = "git:tag:created"
    TAG_DELETED = "git:tag:deleted"

    # Remote repository related
    PUSH = "git:push"
    PULL = "git:pull"
    FETCH = "git:fetch"
    CLONE = "git:clone"

    # Merge related
    MERGE_STARTED = "git:merge:started"
    MERGE_COMPLETED = "git:merge:completed"
    MERGE_CONFLICT = "git:merge:conflict"
    MERGE_ABORTED = "git:merge:aborted"

    # Stash related
    STASH_CREATED = "git:stash:created"
    STASH_APPLIED = "git:stash:applied"
    STASH_DROPPED = "git:stash:dropped"

    # Rebase related
    REBASE_STARTED = "git:rebase:started"
    REBASE_COMPLETED = "git:rebase:completed"
    REBASE_CONFLICT = "git:rebase:conflict"
    REBASE_ABORTED = "git:rebase:aborted"


class BranchType(str, Enum):
    """Git branch type patterns."""

    FEATURE = "feature"
    BUGFIX = "bugfix"
    HOTFIX = "hotfix"
    RELEASE = "release"
    MAIN = "main"
    DEVELOP = "develop"
    OTHER = "other"


class GitAuthor(BaseModel):
    """Git author/committer information."""

    name: str
    email: str
    date: datetime


class GitCommitInfo(BaseModel):
    """Git commit information."""

    hash: str
    short_hash: str
    message: str
    author: GitAuthor
    committer: GitAuthor | None = None
    parents: list[str] = Field(default_factory=list)
    files_changed: int | None = None
    insertions: int | None = None
    deletions: int | None = None
    type: str | None = None  # Conventional Commits type
    scope: str | None = None  # Conventional Commits scope
    breaking: bool = False


class TrackingInfo(BaseModel):
    """Branch tracking information (ahead/behind)."""

    ahead: int = 0
    behind: int = 0


class GitBranchInfo(BaseModel):
    """Git branch information."""

    name: str
    is_remote: bool = False
    is_current: bool = False
    upstream: str | None = None
    last_commit_hash: str | None = None
    last_commit_message: str | None = None
    last_commit_date: datetime | None = None
    tracking: TrackingInfo | None = None


class GitTagInfo(BaseModel):
    """Git tag information."""

    name: str
    type: Literal["lightweight", "annotated"]
    commit_hash: str
    message: str | None = None
    tagger: GitAuthor | None = None


class GitRemoteInfo(BaseModel):
    """Git remote repository information."""

    name: str
    url: str
    type: Literal["fetch", "push"]


class ConflictDetail(BaseModel):
    """Conflict detail for a single file."""

    file: str
    conflict_type: Literal[
        "both-modified",
        "deleted-by-them",
        "deleted-by-us",
        "added-by-them",
        "added-by-us",
    ]


class GitConflictInfo(BaseModel):
    """Git conflict information."""

    files: list[str]
    type: Literal["merge", "rebase", "cherry-pick"]
    source_branch: str | None = None
    target_branch: str | None = None
    details: list[ConflictDetail] = Field(default_factory=list)


class CommitFileChange(BaseModel):
    """File change in a commit."""

    path: str
    status: Literal["added", "modified", "deleted", "renamed"]
    additions: int = 0
    deletions: int = 0


class CommitAnalysis(BaseModel):
    """Commit pattern analysis."""

    is_feature: bool = False
    is_bugfix: bool = False
    is_refactor: bool = False
    is_chore: bool = False
    is_docs: bool = False
    is_test: bool = False


class GitCommitEventData(BaseModel):
    """Git commit event data."""

    commit: GitCommitInfo
    branch: GitBranchInfo
    files: list[CommitFileChange] = Field(default_factory=list)
    analysis: CommitAnalysis | None = None


class MergeInfo(BaseModel):
    """Merge operation information."""

    commits: int
    files: int
    strategy: Literal["fast-forward", "recursive", "octopus", "ours", "subtree"]


class GitBranchEventData(BaseModel):
    """Git branch event data."""

    branch: GitBranchInfo
    previous_branch: str | None = None
    source_branch: str | None = None
    merge_info: MergeInfo | None = None


class GitSyncEventData(BaseModel):
    """Git push/pull/fetch event data."""

    remote: GitRemoteInfo
    branch: str
    commits: int | None = None
    bytes: int | None = None
    direction: Literal["push", "pull", "fetch"]
    success: bool
    error: str | None = None


class GitEvent(BaseEvent):
    """
    Git activity event.

    Represents a Git operation event with detailed information
    about the operation type and affected resources.

    Attributes:
        type: Git event type.
        category: Always EventCategory.GIT.
        data: Git event data payload.
    """

    type: str  # GitEventType value
    category: EventCategory = EventCategory.GIT
    data: dict[str, Any] = Field(default_factory=dict)


class GitEventBuilder:
    """Helper class for creating Git events."""

    @staticmethod
    def create_commit_event(
        event_type: GitEventType,
        data: GitCommitEventData,
        severity: EventSeverity = EventSeverity.INFO,
        correlation_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> GitEvent:
        """
        Create a commit event.

        Args:
            event_type: Type of git event.
            data: Commit event data.
            severity: Event severity level.
            correlation_id: Optional correlation ID.
            metadata: Optional metadata.

        Returns:
            GitEvent instance.
        """
        event_metadata = None
        if metadata:
            event_metadata = EventMetadata(**metadata)

        return GitEvent(
            type=event_type.value,
            category=EventCategory.GIT,
            severity=severity,
            source="GitMonitor",
            data=data.model_dump(),
            correlation_id=correlation_id,
            metadata=event_metadata,
        )

    @staticmethod
    def create_branch_event(
        event_type: GitEventType,
        data: GitBranchEventData,
        severity: EventSeverity = EventSeverity.INFO,
        correlation_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> GitEvent:
        """
        Create a branch event.

        Args:
            event_type: Type of git event.
            data: Branch event data.
            severity: Event severity level.
            correlation_id: Optional correlation ID.
            metadata: Optional metadata.

        Returns:
            GitEvent instance.
        """
        event_metadata = None
        if metadata:
            event_metadata = EventMetadata(**metadata)

        return GitEvent(
            type=event_type.value,
            category=EventCategory.GIT,
            severity=severity,
            source="GitMonitor",
            data=data.model_dump(),
            correlation_id=correlation_id,
            metadata=event_metadata,
        )

    @staticmethod
    def create_sync_event(
        event_type: GitEventType,
        data: GitSyncEventData,
        correlation_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> GitEvent:
        """
        Create a sync (push/pull/fetch) event.

        Args:
            event_type: Type of git event.
            data: Sync event data.
            correlation_id: Optional correlation ID.
            metadata: Optional metadata.

        Returns:
            GitEvent instance.
        """
        event_metadata = None
        if metadata:
            event_metadata = EventMetadata(**metadata)

        severity = EventSeverity.INFO if data.success else EventSeverity.ERROR

        return GitEvent(
            type=event_type.value,
            category=EventCategory.GIT,
            severity=severity,
            source="GitMonitor",
            data=data.model_dump(),
            correlation_id=correlation_id,
            metadata=event_metadata,
        )

    @staticmethod
    def create_conflict_event(
        event_type: GitEventType,
        data: GitConflictInfo,
        correlation_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> GitEvent:
        """
        Create a conflict event.

        Args:
            event_type: Type of git event.
            data: Conflict information.
            correlation_id: Optional correlation ID.
            metadata: Optional metadata.

        Returns:
            GitEvent instance.
        """
        event_metadata = None
        if metadata:
            event_metadata = EventMetadata(**metadata)

        return GitEvent(
            type=event_type.value,
            category=EventCategory.GIT,
            severity=EventSeverity.WARNING,
            source="GitMonitor",
            data=data.model_dump(),
            correlation_id=correlation_id,
            metadata=event_metadata,
        )


def is_git_event(event: BaseEvent) -> bool:
    """Check if event is a git event."""
    return event.category == EventCategory.GIT


def is_git_commit_event(event: BaseEvent) -> bool:
    """Check if event is a git commit event."""
    return event.category == EventCategory.GIT and event.type in (
        GitEventType.COMMIT_CREATED.value,
        GitEventType.COMMIT_AMENDED.value,
    )


def is_git_branch_event(event: BaseEvent) -> bool:
    """Check if event is a git branch event."""
    return (
        event.category == EventCategory.GIT
        and event.type.startswith("git:branch:")
    )


def is_git_sync_event(event: BaseEvent) -> bool:
    """Check if event is a git sync event."""
    return event.category == EventCategory.GIT and event.type in (
        GitEventType.PUSH.value,
        GitEventType.PULL.value,
        GitEventType.FETCH.value,
        GitEventType.CLONE.value,
    )
